File: src/main/java/pw/vodes/rimuru/audit/CommandStaffAction.java
package pw.vodes.rimuru.audit;

import pw.vodes.rimuru.Main;

public class CommandStaffAction {
	
	public String affectedUser, executingUser, reason = "";
	public long time;
	public StaffActionType type;
	
	public CommandStaffAction() {}
	
	public CommandStaffAction(long time, String affectedUser, String executingUser, String reason, StaffActionType type) {
		this.affectedUser = affectedUser;
		this.executingUser = executingUser;
		this.time = time;
		this.type = type;
		this.reason = reason;
	}
	
	@Override
	public boolean equals(Object o) {
		if(o instanceof CommandStaffAction) {
			var csa = (CommandStaffAction)o;
			var diff = Math.abs(time - csa.time);
			if(csa.executingUser.equalsIgnoreCase(Main.api.getYourself().getIdAsString())) {
				return affectedUser.equalsIgnoreCase(csa.affectedUser)
						&& diff < 3
						&& type == csa.type;
			} else {
				return affectedUser.equalsIgnoreCase(csa.affectedUser)
						&& executingUser.equals(csa.executingUser)
						&& diff < 3
						&& type == csa.type
						&& reason.equalsIgnoreCase(csa.reason);
			}
		}
		return super.equals(o);
	}

}

File: src/main/kotlin/pw/vodes/rimurukt/audit/CommandStaffAction.kt
package pw.vodes.rimurukt.audit

import kotlinx.serialization.Serializable
import org.javacord.api.entity.message.embed.EmbedBuilder
import pw.vodes.rimurukt.Main
import kotlin.math.abs

@Serializable
data class StaffAction(val affectedUser: String, val executingUser: String, val time: Long, val type: StaffActionType, val reason: String = "") {

    override fun equals(other: Any?): Boolean {
        if (other !is StaffAction)
            return super.equals(other)

        val diff = abs(time - other.time)

        return if (other.executingUser.equals(Main.api.yourself.idAsString, true)) {
            affectedUser.equals(other.affectedUser, true) &&
                    diff < 4 &&
                    type == other.type
        } else
            affectedUser.equals(other.affectedUser, true) &&
                    executingUser.equals(other.executingUser, true) &&
                    diff < 4 &&
                    type == other.type
    }

    fun getEmbed(): EmbedBuilder {
        val u1 = Main.api.getUserById(executingUser).get()
        val u2 = Main.api.getUserById(affectedUser).get()

        return EmbedBuilder().setTitle(type.messageTitle()).setFooter("ID: $affectedUser").setAuthor(u1).setThumbnail(u2.getAvatar(4096))
            .setDescription(u2.name + if (reason.isBlank()) "" else "\nReason:\n```$reason```")
    }
}
